software/database/src/main/java/brooklyn/entity/database/mysql/MySqlClusterImpl.java

```diff
@@ -0,0 +1,445 @@
+            String masterLogFile = validateSqlParam(getAttributeBlocking(master, MySqlMaster.MASTER_LOG_FILE));
+            Integer masterLogPos = getAttributeBlocking(master, MySqlMaster.MASTER_LOG_POSITION);
+            String masterAddress = validateSqlParam(master.getAttribute(MySqlNode.SUBNET_ADDRESS));
+            Integer masterPort = master.getAttribute(MySqlNode.MYSQL_PORT);
+            String slaveAddress = validateSqlParam(slave.getAttribute(MySqlNode.SUBNET_ADDRESS));
+            String username = validateSqlParam(cluster.getConfig(SLAVE_USERNAME));
+            String password = validateSqlParam(cluster.getAttribute(SLAVE_PASSWORD));
+
+            executeScriptOnNode(master, String.format(
+                    "CREATE USER '%s'@'%s' IDENTIFIED BY '%s';\n" +
+                    "GRANT REPLICATION SLAVE ON *.* TO '%s'@'%s';\n",
+                    username, slaveAddress, password, username, slaveAddress));
+
+            String slaveCmd = String.format(
+                    "CHANGE MASTER TO " +
+                        "MASTER_HOST='%s', " +
+                        "MASTER_PORT=%d, " +
+                        "MASTER_USER='%s', " +
+                        "MASTER_PASSWORD='%s', " +
+                        "MASTER_LOG_FILE='%s', " +
+                        "MASTER_LOG_POS=%d;\n" +
+                    "START SLAVE;\n",
+                    masterAddress, masterPort, username, password, masterLogFile, masterLogPos);
+            executeScriptOnNode(slave, slaveCmd);
+
+            cluster.getAttribute(SLAVE_ID_ADDRESS_MAPPING).put(slave.getId(), slave.getAttribute(MySqlNode.SUBNET_ADDRESS));
+        }
+
+        private <T> T getAttributeBlocking(Entity masterNode, AttributeSensor<T> att) {
+            return DynamicTasks.queue(DependentConfiguration.attributeWhenReady(masterNode, att)).getUnchecked();
+        }
+
+    }
+
+    private static final class NodeRunningListener implements SensorEventListener<Boolean> {
+        private MySqlCluster cluster;
+
+        public NodeRunningListener(MySqlCluster cluster) {
+            this.cluster = cluster;
+        }
+
+        @Override
+        public void onEvent(SensorEvent<Boolean> event) {
+            final MySqlNode node = (MySqlNode) event.getSource();
+            if (Boolean.TRUE.equals(event.getValue()) &&
+                    // We are interested in SERVICE_PROCESS_IS_RUNNING only while haven't come online yet.
+                    // Probably will get several updates while replication is initialized so an additional
+                    // check is needed whether we have already seen this.
+                    Boolean.FALSE.equals(node.getAttribute(MySqlNode.SERVICE_UP)) &&
+                    !Boolean.TRUE.equals(node.getAttribute(NODE_REPLICATION_INITIALIZED))) {
+
+                // Events executed sequentially so no need to synchronize here.
+                if (Boolean.TRUE.equals(node.getAttribute(NODE_REPLICATION_INITIALIZED))) {
```

Comment:
Shouldn't this condition never be true given the opposite condition in the if-statement above? If so then it can be removed.